File: 930. Binary Subarrays With Sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        Map<Integer, Integer> prefixCount = new HashMap<>();
        int sum = 0, count = 0;
        prefixCount.put(0, 1);
        for (int i : nums) {
            sum += i;
            count += prefixCount.getOrDefault(sum - goal, 0);
            prefixCount.put(sum, prefixCount.getOrDefault(sum, 0) + 1);
        }
        return count;
    }
}

File: go/930. Binary Subarrays With Sum.go
func numSubarraysWithSum(nums []int, goal int) int {
    prefixCount := map[int]int{}
    sum, count := 0, 0
    prefixCount[0] = 1
    for _, i := range(nums) {
        sum += i
        count += prefixCount[sum - goal]
        prefixCount[sum] += 1
    }
    return count
}

